Add root_nodes parameter to ExpertKnowledge

MAVRICK-1 opened the pull request:
- Added root_nodes parameter to ExpertKnowledge.__init__
- Implemented _add_root_nodes_forbidden_edges() to automatically convert root nodes to forbidden edges
- Updated apply_expert_knowledge() and limit_search_space() to process root nodes
- Added comprehensive documentation and examples
- Added test_root_nodes_single() and test_root_nodes_multiple() tests


Fixes https://github.com/pgmpy/pgmpy/issues/2407

MAVRICK-1 committed:
Add root_nodes parameter to ExpertKnowledge

- Added root_nodes parameter to ExpertKnowledge.__init__
- Implemented _add_root_nodes_forbidden_edges() to automatically convert root nodes to forbidden edges
- Updated apply_expert_knowledge() and limit_search_space() to process root nodes
- Added comprehensive documentation and examples
- Added test_root_nodes_single() and test_root_nodes_multiple() tests
- Fixed flake8 E721 error (isinstance instead of type comparison)

Fixes #2407

MAVRICK-1 committed:
refactor: remove unused import os from test_expert.py

MAVRICK-1 commented:
cc @ankurankan

ankurankan commented:
@MAVRICK-1 Thanks for the PR. The tests seem to be failing, could you please have a look?

MAVRICK-1 commented:
Its minor issue solving it

MAVRICK-1 committed:
feat: add support for root node forbidden edges in ExpertInLoop and update tests

MAVRICK-1 committed:
refactor: simplify dictionary comprehension in TestExpertInLoop

MAVRICK-1 committed:
Merge branch 'dev' into feature/add-root-nodes-expert-knowledge

MAVRICK-1 commented:
@ankurankan all test passed

ankurankan commented on pgmpy/estimators/ExpertKnowledge.py:
```diff
@@ -290,3 +337,6 @@ def limit_search_space(self, data_coulumn_labels):
         self.forbidden_edges = self.forbidden_edges.union(forbidden_edges_additive)
+
+        # Also add forbidden edges for root nodes
+        self._add_root_nodes_forbidden_edges(data_coulumn_labels)
```
I think this needs to be called in the `__init__` method itself. There are other causal discovery algorithms that use expert knowledge, for example, PC, Hill-Climb, GES, etc.

ankurankan commented on pgmpy/tests/test_estimators/test_expert.py:
```diff
@@ -272,8 +278,77 @@ def test_edge_orientation_priority(self):
         if ("Income", "Education") in dag.edges():
             assert ("Education", "Income") not in dag.edges()
+
+    @unittest.skipUnless(
```
Could you also please add tests for the `root_node` in some other causal discovery algorithms like PC, Hill-Climb Search ?

MAVRICK-1 committed:
Merge branch 'dev' into feature/add-root-nodes-expert-knowledge

MAVRICK-1 commented:
@ankurankan I will resolve everything in this PR, as well as the others, by the end of the week. My exams are currently going on.

ankurankan committed:
Merge branch 'dev' into feature/add-root-nodes-expert-knowledge

MAVRICK-1 committed:
refactor(estimators): Centralize root node processing in ExpertKnowledge

This commit refactors the  class to centralize the processing of .

Previously, the logic to add forbidden edges for root nodes was called from multiple places within different structure learning algorithms (, , , , ). This led to inconsistent behavior, especially when  were specified without a .

This commit introduces a new method  to the  class. This method is now the single point of entry for processing  and is called by all structure learning algorithms that use . This ensures that root node constraints are applied consistently across all algorithms.

The following changes were made:
- Added  method to  to handle root node logic.
- Removed calls to  from  and  in .
- Updated , , , , and  estimators to call  at the beginning of their  methods.
- Fixed a minor typo in .
- Added tests for  in , , , and .

MAVRICK-1 committed:
refactor(estimators): Centralize root node processing in ExpertKnowledge

This commit refactors the  class to centralize the processing of .

Previously, the logic to add forbidden edges for root nodes was called from multiple places within different structure learning algorithms (, , ). This led to inconsistent behavior, especially when  were specified without a .

This commit introduces a new method  to the  class. This method is now the single point of entry for processing  and is called by all structure learning algorithms that use . This ensures that root node constraints are applied consistently across all algorithms.

The following changes were made:
- Added  method to  to handle root node logic.
- Removed calls to  from  and  in .
- Updated , , and  estimators to call  at the beginning of their  methods.
- Fixed a minor typo in .

MAVRICK-1 committed:
Merge branch 'dev' into feature/add-root-nodes-expert-knowledge

MAVRICK-1 commented:
@ankurankan  I've tidied up the root node handling in ExpertKnowledge - basically consolidated everything into one set_variables method so it works the same way across all our algorithms (PC, HillClimbSearch, GES, MmhcEstimator, ExpertInLoop). Added some tests to make sure it all works properly.

MAVRICK-1 committed:
refactor: Clean up imports in MmhcEstimator and test_root_node_algorithms

MAVRICK-1 committed:
refactor(estimators): Remove redundant forbidden edges handling for root nodes

MAVRICK-1 commented:
@ankurankan all the test are passing

MAVRICK-1 commented:
@ankurankan any more changes needed in this pr?

MAVRICK-1 commented:
Cc @ankurankan

ankurankan commented:
@Nimish-4 Would you be interested in reviewing this?

MAVRICK-1 committed:
Merge branch 'dev' into feature/add-root-nodes-expert-knowledge

MAVRICK-1 commented:
resloving the conflict

ankurankan commented:
Closing this as the changes are not relevant anymore. Needs to be implemented in the new `causal_discovery` module
